Add terminology section to the buildpack spec (#300)
buildpacks/spec: buildpack.md

 - `<major>` and `<minor>` MUST only contain numbers (unsigned 64 bit integer)
 - When `<major>` is greater than `0` increments to `<minor>` SHALL exclusively indicate additive changes

## Terminology

> ### CNB Terminology

Author: Feedback from 4/21 Working Group: (if possible, without making each term its own section) we should link back to this section as much as is practical.

> A **buildpack** is a directory containing a `buildpack.toml`.

Reviewer: While technically true, we probably want a definition to also describe what a buildpack does?

```suggestion
A **buildpack** is a directory containing a `buildpack.toml`.
```

Author: Updated in ff54426

A **buildpack group**, or **group**, is a list of one or more buildpacks that are designed to work together - for example, a buildpack that provides `node` and a buildpack that provides `npm`.

An **order** is a list of one or more groups to be tested against application source code, so that the appropriate group for a build can be determined. 

> An **executable buildpack** is a buildpack containing `/bin/detect` and `/bin/build` executables.

Author: ```suggestion
A **component buildpack** is a buildpack containing `/bin/detect` and `/bin/build` executables.
```
Feedback from 4/21 Working Group: `executable buildpack` is a little misleading because composite buildpacks are still "executable" in the sense that their components are executable.

> An **order buildpack** is a buildpack containing an order definition in `buildpack.toml`. Order buildpacks do not contain `/bin/detect` or `/bin/build` executables.

Author: Suggestions from 4/21 Working Group: 
* composite buildpack (winner)
* multi buildpack
* anything but meta buildpack

> An **order buildpack** is a buildpack containing an order definition in `buildpack.toml`. Order buildpacks do not contain `/bin/detect` or `/bin/build` executables.

Author: ```suggestion
An **composite buildpack** is a buildpack containing an order definition in `buildpack.toml`. Composite buildpacks do not contain `/bin/detect` or `/bin/build` executables.
```

**Resolving an order** is the process by which an order (which may contain order buildpacks) is evaluated together with application source code to produce a group of executable buildpacks that can be used to build the application. This process is known as **detection**. During detection, the `/bin/detect` executable for each executable buildpack is invoked, and the group that is returned is the first group where `/bin/detect` returns true for all non-optional executable buildpacks.

A **lifecycle** is software that orchestrates a build. It executes in a series of phases that each have a distinct responsibility:
  1. **Detection,** where an optimal selection of compatible buildpacks is chosen by resolving the provided order.
  2. **Analysis,** where metadata about OCI layers generated during a previous build are made available to executable buildpacks in the selected group.
  3. **Build,** where the `bin/build` for each executable buildpack in the selected group is invoked, in order, to build the application.
  4. **Export,** where filesystem changes from the build phase are packaged into layers in an OCI image.

> A **launcher** is software contributed by the lifecycle as the `ENTRYPOINT` of the exported OCI image that is used to start processes at runtime.

Reviewer: Do we want to add other things here that a launcher does like load the environment or is that intentionally left off?

Author: Update in ff54426

**Launch** describes the process of running an application by creating a container from the exported OCI image.

A **platform** is a system or software that orchestrates the lifecycle by invoking each lifecycle phase in order.

> A **process type** is a definition, provided by executable buildpacks during the build phase, of a process to launch at runtime.

Author: Feedback from 4/21 Working Group: we should change `process type` to `process definition` and only use `process type` when we are talking about the specific key `type`.

> A **process type** is a definition, provided by executable buildpacks during the build phase, of a process to launch at runtime.

Author: ```suggestion
A **process definition** is a definition, provided by executable buildpacks during the build phase, of a process to launch at runtime.
```

Author: "definition is a definition" is a bit weird... I can't think how to improve it.

#### Order Resolution

During detection, an order definition MUST be resolved into individual component buildpacks.

The resolution process MUST follow this pattern:

Where:
- O and P are buildpack orders.

> - A through H are component buildpacks.

Author: I think this is correct. But I'm a bit less sure of myself here.

   - `id` MUST be set to the special value `"*"`.
   - `mixins` MUST be empty.

#### Order Buildpacks

> A buildpack descriptor that specifies `order` MUST be [resolvable](#order-resolution) into an ordering of buildpacks that implement the [Buildpack Interface](#buildpack-interface).

Author: Moved to terminology
